### PR TITLE
Fix/async issues on badge creation page

### PR DESCRIPTION
**Pull request by timber-they**
This *should* solve #628 and #480. However, someone will need to test it on Safari, since I can't do this on Linux and some errors only appeared there.

I removed a lot of the stuff that made it complicated and now it works, but I have the suspicion, that some of it was necessary to make it work on Safari. However, I do think that it should be possible to get it to work in Safari with this simpler approach as well. So please do give me feedback if / what exactly fails.

### CODE REVIEW

**Review by YouQam — COMMENTED** [review]
The issue with safari (not showing image-frame from first time) is back again in this PR.
See attached video.

https://github.com/user-attachments/assets/9707c295-8dc6-47e7-addb-4d0bc431aa6d

**Comment by timber-they** [comment]
@YouQam are there any logs indicating an error or something similar?

**Comment by YouQam** [comment]
> @YouQam are there any logs indicating an error or something similar?

Yes, this is the error raised on safari browser-console 
Screenshot 2024-08-29 at 13 25 57

File: src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html [hunk 1]
```diff
@@ -1,3 +1,10 @@
+<!-- loading spinner -->
```
Comment (timber-they) [inline src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html:1]: I think this is from another ticket

File: src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ export class ExportPdfDialog extends BaseDialog {
 
 	profile: UserProfile;
 	emailsLoaded: Promise<unknown>;
+	pdfIsLoading: boolean = false;
```
Comment (timber-they) [inline src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.ts:33]: This is also from another ticket

File: src/app/issuer/components/badge-studio/badge-studio.component.ts [hunk 3]
```diff
@@ -233,10 +233,12 @@ export class BadgeStudioComponent implements OnInit, OnChanges {
 							'data:image/svg+xml;charset=utf-8,' +
 							encodeURIComponent(new XMLSerializer().serializeToString(svgRoot));
 						// const svgDataUrl2 = 'data:image/svg+xml;charset=utf-8,' + encodeURIComponent(new XMLSerializer().serializeToString(svgRoot2));
-						const svgImage = new Image();
-						const svgImage2 = new Image();
+						let svgImage = new Image();
```
Comment (timber-they) [inline src/app/issuer/components/badge-studio/badge-studio.component.ts:236]: Why is it necessary to make this mutable? You should be able to reassign properties / attributes whilst it being `const`

File: src/app/issuer/components/badge-studio/badge-studio.component.ts [hunk 4]
```diff
@@ -278,8 +280,17 @@ export class BadgeStudioComponent implements OnInit, OnChanges {
...
 						};
+
 						svgImage.src = svgDataUrl;
-						svgImage2.src = this.imageToShow;
+
+						// To resolve the issue of missing frame, we need to make sure that frame-image is aleady loaded before drawing badge-image
+						if(svgImage.complete){
+							svgImage2.src = this.imageToShow;
+						} else {
+							setTimeout(() => {
```
Comment (timber-they) [inline src/app/issuer/components/badge-studio/badge-studio.component.ts:290]: All those `setTimeout` solutions are very messy. I'll let it pass if you think it's the best solution, but it'd be a lot cleaner not have to use them

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 5]
```diff
@@ -284,17 +284,15 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 	// To get and download badge certificate in pdf format
 	downloadCertificate(instance: BadgeInstance, badgeIndex: number) {
 		this.downloadStates[badgeIndex] = true;
-		this.pdfService.getPdf(instance.slug).subscribe(
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:287]: This is also from another ticket

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 6]
```diff
@@ -499,10 +479,21 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
...
-		// by regenerating the upload image as adjusting upload image didn't work with this issue
-		if (this.platformService.SAFARI && this.currentImage) {
-			this.generateUploadImage(this.currentImage, this.badgeClassForm.value);
+		// To update badge-frame when badge-category is changed
+		this.changeBadgeFrame();
+	}
+
+	changeBadgeFrame() {
+		if (this.imageField?.control.value) {
+			setTimeout(() => {
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:488]: Same here. Timeouts like this shouldn't be necessary. There *should* exist a cleaner solution to this problem

File: src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html (hunk 1 above)
Comment (YouQam) [reply src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html:1]: You are right! I wanted to test st by adding it here.

File: src/app/issuer/components/badge-studio/badge-studio.component.ts (hunk 3 above)
Comment (YouQam) [reply src/app/issuer/components/badge-studio/badge-studio.component.ts:236]: I made it `constant` again.

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 6 above)
Comment (YouQam) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:488]: It doesn't work without adding `timeout`. Dealing with image loading seems to be tricky sepcially with safari browser.

File: src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html (hunk 1 above)
Comment (YouQam) [reply src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.html]: reverted.

File: src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.ts (hunk 2 above)
Comment (YouQam) [reply src/app/common/dialogs/export-pdf-dialog/export-pdf-dialog.component.ts]: reverted.

File: src/app/common/services/pdf.service.ts [hunk 7]
```diff
@@ -6,7 +6,7 @@ import { MessageService } from './message.service';
 import { SessionService } from './session.service';
 import { HttpHeaders, HttpResponse } from '@angular/common/http';
 import { DomSanitizer, SafeResourceUrl } from '@angular/platform-browser';
-import { Observable, map } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (YouQam) [reply src/app/common/services/pdf.service.ts]: reverted.

File: src/app/issuer/components/badge-studio/badge-studio.component.ts (hunk 4 above)
Comment (YouQam) [reply src/app/issuer/components/badge-studio/badge-studio.component.ts:290]: I think this is the best solution for now.

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts (hunk 5 above)
Comment (YouQam) [reply src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts]: reverted.

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 6 above)
Comment (YouQam) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts]: I removed them after updating the code.

**Review by timber-they — COMMENTED** [review]
I can't approve this PR because I created it, but I *would* approve it.

**Comment by YouQam** [comment]
> I can't approve this PR because I created it, but I _would_ approve it.

@zven Can you test this PR out using safari then approve?

**Comment by zven** [comment]
@YouQam I found at least one issue:
* tested on Safari
* cach is cleaned and website data removed
* open badge creation page
* select "Teilnahme"
* click "aus bestehenden Icons wählen" and pick icon -> icon is selected and properly shown (with frame around)
* click "andere Datei auswählen" and upload file (see `beige.png` below)-> previous imag is still shown, even after multiple attempts of uploading int

Below find the image I used for testing:
beige

**Comment by zven** [comment]
Update: Similar behaviour when you want to pick another icon from the noun project -> old icon still displayed

**Comment by zven** [comment]
When editing a badge:
* open existing badge for editing
* pick another icon via "anderes Icon suchen" -> old icon still displayed

**Comment by zven** [comment]
@YouQam The issue mentioned above is now indeed fixed. 👍

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 8]
```diff
@@ -1022,23 +1018,24 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
...
 	}
 
-	adjustUploadImage(formdata) {
+	/**
+	 * Updates image-frame when category is changed.
+	 * 
+	 * @param formdata
+	 * @param isCategoryChanged - To prevent unnecessary call of (@function generateUploadImage) which causes an issue of drawing multiple frames around badge-image.
+	 */
+	UpdateImageFrame(formdata, isCategoryChanged=false) {
```
Comment (jona159) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:1033]: Please dont use uppercase for function names

**Review by jona159 — CHANGES_REQUESTED** [review]
Please change function name to lower case

**Review by jona159 — APPROVED** [review]

**Pull request merged**